netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -268,43 +240,42 @@ protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
    * Attempts to create a new stream from the given command. If there are too many active streams,
    * the creation request is queued.
    */
-  private void createStream(CreateStreamCommand command, ChannelPromise promise) {
-    // Add the creation request to the queue.
-    pendingStreams.addLast(new PendingStream(command, promise));
-
-    // Process the pending streams queue.
-    createPendingStreams();
+  private void createStream(CreateStreamCommand command, final ChannelPromise promise) {
+    Http2Connection connection = connection();
+    if (connection.isGoAway()) {
+      promise.setFailure(goAwayStatus().asException());
```

Comment:
It seems we need stream.transportReportStatus(goAwayStatus()) here. Before transportReportStatus calling promise.setSuccess() may make more sense because 1) we don't care about the promises and creating the exception is costly, 2) this can be treated identically as if the GOAWAY was received after the HEADERS was sent without changing semantics.